src/backend/TrafficCourts/Citizen.Service/Services/Tickets/Search/Rsi/Authentication/AccessTokenUpdateWorker.cs

```diff
@@ -0,0 +1,123 @@
+                // or established connection failed because connected host has failed to respond.
+
+                // todo: log specific message about not being able to connect, not on VPN or address wrong
+                var rethrow = await OnError(exception);
+                if (rethrow)
+                {
+                    throw;
+                }
+            }
+            catch (HttpRequestException exception)
+            {
+                var rethrow = await OnError(exception);
+                if (rethrow)
+                {
+                    throw;
+                }
+            }
+        }
+
+        async Task<bool> OnError(Exception exception)
+        {
+            var delay = TimeSpan.FromMilliseconds(500 + _random.Next(50, 100));
+            _logger.LogInformation(exception, "Failed to update access token on {Attempt}, will delay {DelayTime} before next attempt", ++retry, delay);
+            await Task.Delay(delay, stoppingToken);
+            return false;
+        }
+    }
+
+    private async Task DoWorkAsync(CancellationToken stoppingToken)
+    {
+        _logger.LogInformation("Updating access token");
+        var nextRefresh = await UpdateTokenAsync(stoppingToken);
+
+        var delay = nextRefresh - DateTimeOffset.UtcNow;
+        if (TimeSpan.Zero < delay)
+        {
+            using (_logger.BeginScope(new Dictionary<string, object> { { "Delay", delay } }))
+            {
+                _logger.LogInformation("Waiting for next access token refresh");
+            }
+
+            await Task.Delay(delay, stoppingToken);
+        }
+
+    }
+
+    /// <summary>
+    /// Gets the new access token and returns the date when the next refresh attempt should 
+    /// occur.
+    /// </summary>
+    /// <param name="cancellationToken"></param>
+    /// <returns></returns>
+    private async Task<DateTimeOffset> UpdateTokenAsync(CancellationToken cancellationToken)
+    {
+        using var activity = Diagnostics.Source.StartActivity("Update Access Token");
+
+        try
+        {
+            var token = await _authenticationClient.GetTokenAsync(cancellationToken);
+            // next refresh is in 1/2 the token life time +/- 30 seconds.
+            var nextRefresh = DateTimeOffset.UtcNow.AddSeconds(token.ExpiresIn / 2 - _random.Next(-30, 30));
+            _cache.SaveToken(token);
+            activity?.SetStatus(System.Diagnostics.ActivityStatusCode.Ok);
+            return nextRefresh;
+        }
+        catch (AuthenticationException ex) when (ex.StatusCode != System.Net.HttpStatusCode.Unauthorized)
+        {
+            // config error?
+            // service down?
+            _logger.LogWarning(ex, "Failed to get new access token");
+            activity?.SetStatus(System.Diagnostics.ActivityStatusCode.Error);
+            return DateTimeOffset.UtcNow.AddSeconds(_random.Next(5, 30)); // try again in 5-30 seconds
```

Comment:
5-30 seconds just seemed reasonable. The access token is typically valid for 5 minutes, and updating at 1/2 that lifetime, means at worst, we would have to fail 5 times at 30 second delay before we would would no longer have a valid access token.  We dont retry if the credentials are incorrect (Unauthorized).